libres/lib/enkf/enkf_main.cpp

```diff
@@ -743,6 +737,337 @@ static FILE *enkf_main_log_step_list(const char *log_path,
+
+        enkf_main_serialize_dataset(ensemble_config, dataset, last_step,
+                                    use_count, tp, serialize_info);
+        for (int i = 0; i < cpu_threads; i++) {
+            serialize_info[i].A = matrix_alloc_copy(A);
+        }
+
+        enkf_main_deserialize_dataset(ensemble_config, dataset, serialize_info,
+                                      tp);
+        serialize_info_free(serialize_info);
+    }
+    thread_pool_free(tp);
+}
+
+static std::unordered_map<
+    std::string,
+    std::vector<std::pair<matrix_type *, const row_scaling_type *>>>
+enkf_main_load_row_scaling_parameters(
+    enkf_fs_type *target_fs, ensemble_config_type *ensemble_config,
+    int_vector_type *iens_active_index, int last_step, run_mode_type run_mode,
+    meas_data_type *forecast, enkf_state_type **ensemble, hash_type *use_count,
+    obs_data_type *obs_data, const local_ministep_type *ministep) {
+
+    int matrix_start_size = 250000;
+    int active_ens_size = meas_data_get_active_ens_size(forecast);
+    matrix_type *A = matrix_alloc(matrix_start_size, active_ens_size);
+
+    std::unordered_map<
+        std::string,
+        std::vector<std::pair<matrix_type *, const row_scaling_type *>>>
+        parameters;
+    for (auto &[dataset_name, dataset] :
+         local_ministep_get_datasets(ministep)) {
+
+        if (!(local_dataset_get_size(dataset)))
+            continue;
+
+        const auto &scaled_keys = local_dataset_scaled_keys(dataset);
+
+        if (scaled_keys.size() == 0)
+            continue;
+        local_obsdata_type *local_obsdata =
+            local_ministep_get_obsdata(ministep);
+
+        std::vector<std::pair<matrix_type *, const row_scaling_type *>>
+            row_scaling_list;
+        for (int ikw = 0; ikw < scaled_keys.size(); ikw++) {
+            const auto &key = scaled_keys[ikw];
+            const active_list_type *active_list =
+                local_dataset_get_node_active_list(dataset, key.c_str());
+            const auto matrix_rows = matrix_get_rows(A);
+            const auto *config_node =
+                ensemble_config_get_node(ensemble_config, key.c_str());
+            const auto node_size =
+                enkf_config_node_get_data_size(config_node, last_step);
+            if (matrix_get_rows(A) < node_size)
+                matrix_resize(A, node_size, active_ens_size, false);
+
+            for (int iens = 0; iens < int_vector_size(iens_active_index);
+                 iens++) {
+                int column = int_vector_iget(iens_active_index, iens);
+                if (column >= 0) {
+                    serialize_node(target_fs, config_node, iens, last_step, 0,
+                                   column, active_list, A);
+                }
+            }
+            const row_scaling_type *row_scaling =
+                local_dataset_get_row_scaling(dataset, key.c_str());
+            matrix_shrink_header(A, row_scaling->size(), matrix_get_columns(A));
```

Comment:
Not really related to this PR, but do you know why we want to just shrink the headers?

Reply:
it is a fast way of resizing the matrix without changing the allocated memory